forge-ai/src/main/java/forge/ai/ability/EffectAi.java

```diff
@@ -256,6 +271,171 @@ public boolean apply(final Card c) {
+                            Map<AbilityKey, Object> runParams = AbilityKey.mapFromAffected(input);
+                            runParams.put(AbilityKey.Regeneration, true);
+                            List<ReplacementEffect> repDestoryList = game.getReplacementHandler().getReplacementList(ReplacementType.Destroy, runParams, ReplacementLayer.Other);
+                            // no Destroy Replacement, or one non-Regeneration one like Totem-Armor
+                            if (repDestoryList.isEmpty() || Iterables.any(repDestoryList, Predicates.not(CardTraitPredicates.hasParam("Regeneration")))) {
+                                return false;
+                            }
+                            // TODO check Stack for Effects that would destroy host, either direct or indirect
+
+                            return true;
+                        }
+                    });
+
+                    if (list.isEmpty()) {
+                        return false;
+                    }
+                    // TODO check Stack for Effects that would destroy the selected card?
+                    sa.getTargets().add(ComputerUtilCard.getBestAI(list));
+                    return true;
+                } else if (sa.getParent() != null) {
+                    // sub ability should be okay
+                    return true;
+                } else if ("Self".equals(sa.getParam("RememberObjects"))) {
+                    // the ones affecting itself are Nimbus cards, were opponent can activate this effect
+                    Card host = sa.getHostCard();
+                    if (!host.canBeDestroyed()) {
+                        return false;
+                    }
+                    // do this only in reaction to a threatening spell on directly on the stack
+                    MagicStack stack = game.getStack();
+                    if (stack.isEmpty()) {
+                        return false;
+                    }
+
+                    Map<AbilityKey, Object> runParams = AbilityKey.mapFromAffected(sa.getHostCard());
+                    runParams.put(AbilityKey.Regeneration, true);
+                    List<ReplacementEffect> repDestoryList = game.getReplacementHandler().getReplacementList(ReplacementType.Destroy, runParams, ReplacementLayer.Other);
+                    // no Destroy Replacement, or one non-Regeneration one like Totem-Armor
+                    if (repDestoryList.isEmpty() || Iterables.any(repDestoryList, Predicates.not(CardTraitPredicates.hasParam("Regeneration")))) {
+                        return false;
+                    }
+                    // TODO check Stack for Effects that would destroy host, either direct or indirect
+                    SpellAbility stackSa = stack.peekAbility();
+                    if (stackSa == null) {
+                        return false;
+                    }
+
+                    // need to get the wrapped sub abilities
+                    if (stackSa.isWrapper()) {
```

Comment:
imo all calls below should be handled by WrappedAbility calls
so no need to extra unwrap?

Reply:
hm maybe, you could remove it if you want